### DISCUSSION TITLE
UI display for equippable resources

### DISCUSSION THREAD

**Discussion by Erebos0506** [rmrk-team/rmrk-spec#51] [Ideas]
### Proposal: 
Adding a way to hide the main resource("src" field) (and show thumb instead) for equippable nfts/resources when viewing the item

### Reasons:
- As some equippables only make sense in conjunction with the base composed image showing the actual equippable resource would lead to weird looking images that don't make sense by themselves or just look unfinished thereby negatively affecting user experience 
- Not showing the resource that will be shown after equipping would make it possible/easier to create puzzles/eastereggs or other nfts where the owners are encouraged to explore the different possibilities offered on their own instead of being presented the solution by looking at the item
- In some cases a item might even just be used to hide a fallback leading to the item not showing anything as a transparent picture is used as resource causing users to think there are some problems with the items when they're actually working fine

### Possible solutions:
- Adding a field to the nft metadata to specify whether to show the "src" or "thumb" field for items - eg `"concealed":"true"`
- Instead of adding a field with boolean value add a field with ipfs link to another picture that will be displayed instead thereby enabling higher quality pictures to be shown when viewing the item instead of the thumb - eg `"substitute":"ipfs://ipfs/HASH"`

### Possible problems:
- Buyers won't see exactly what they're buying -> Possible partial solution: Add a UI warning that says that the equippable resource isn't shown when viewing the item
- Would need the UI to read from the metadata which resource to show which might slow down loading a bit as metadata would need to be loaded before the images can start to be loaded -> The loading time shouldn't be affected much especially as only items are affected by this

**Comment by MaEdAlpha** [comment]
Just an idea and some psuedo code.
when minting, for  <Resource> type add a displayThumbAsItem:boolean.


```
          resaddSendRemarks.push(
          equipableNft.resadd({
            //adds resource link between slot and equipableNft.
            src: `ipfs://ipfs/${ASSETS_CID}/Wegland%20Items/${resource}`,
            thumb: `ipfs://ipfs/${ASSETS_CID}/Wegland%20Items/${item.thumb}`,
            id: primaryItemResource,
            slot: slotName,
            displayThumbAsItem: true,
          })
        );
```
        
 Then on front end client side, 
 
 toggle your img path based off this boolean.
 
` <svg src= { resource.displayThumbAsItem ? resource.thumb : resource.src } >`

**Comment by Yuripetusko** [comment]
sounds easy enough. I would perhaps use a different name. `preferThumb`

**Reply by MaEdAlpha** [reply]
nft.d.ts file.
```
export interface Resource {
    id?: string;
    base?: string;
    src?: string;
    parts?: string[];
    metadata?: string;
    slot?: string;
    pending?: boolean;
    thumb?: string;
    theme?: Theme;
    themeId?: string;
    preferThumb?:boolean
}
```

nft.js file. 
```
export class NFT {
    constructor(nftInstance) {
        this.changes = [];
        //.....
        this.preferThumb = false;
    }
```

**Reply by MaEdAlpha** [reply]
default to false, that way nothing really changes for current system.  Put a lil ternary in on the front-end and good to go.

**Reply by Yuripetusko** [reply]
I actually think it's best to add it to metadata of a resource on ipfs. in properties

**Reply by MaEdAlpha** [reply]
Kind of like what RMRK did with `IRoyaltyAttribute`?

```
export interface IPreferredImage extends IAttribute {
    type: "item-display";
    value: {
        overRide: boolean;
        thumb: string;
    };
}
```

**Reply by MaEdAlpha** [reply]
overRide -> lets you use this image  if true, ignores if false (maybe a toggle feature later on dev side)
thumb -> ipfs string.

**Reply by Yuripetusko** [reply]
No no, I just mean that it's better to put it on IPFS than on chain as a new resource field.

It can be part of metadata spec. preferThumb

**Reply by MaEdAlpha** [reply]
> No no, I just mean that it's better to put it on IPFS than on chain as a new resource field.
> 
> It can be part of metadata spec. preferThumb

ahhh I gotcha.  so something like this?


```
  const metadataCid = await pinSingleMetadataFromDir(
        `${ASSET_PATH}/Wegland Items`,
        item.thumb,
        item.name,
        {
          description: item.description,
          image: `.....`,
          externalUri: `.....`
          preferThumb:   `ipfs://ipfs/${ASSETS_CID}/PathTo-UI-Default`
        }
      )
```

**Reply by Yuripetusko** [reply]
well kinda but it should be ipfs cid basically already pinned., and I think it can go on properties on metadata on ipfs. 

I think I'll make a PR in spec today

**Reply by Yuripetusko** [reply]
https://github.com/rmrk-team/rmrk-spec/pull/52/files

**Reply by MaEdAlpha** [reply]
sorry for the delayed response. 

essentially when pinning the metadataCID  IPFS, our TS interface should have this added in. 

![image](https://user-images.githubusercontent.com/72553021/167238762-9f82e5cf-aa7e-4f01-a374-718b0c4db975.png)

**Comment by MaEdAlpha** [comment]
Ahhh I see, sorry that part didn't sink in. I was mostly responding to my current need which is a low res 'default' image. High res artwork may not work because pngs can get 'fuzzy' based off of screen dimensions, and probably shouldn't be so large...or it defeats the purpose.
